### PR TITLE
Rework C++ implementation for Huffman encoding

### PR DESCRIPTION
**Pull request by antonte**
- Make implementation shorter and easier to understand
- Make namespace hierarchy flat
- Combine Leaf and Branch in one class Node for the tree
- Make codebook explicit
- Use std::string for string_to_encode
- Remove noise (delete function with_new_bit() for instance)
- Use unordered map for weights instead of vector
- Instead inserting to the orphan array, sort it on every iteration (debatable)
- Remove structure for cacheing bits for character, calculate it every time
- Remove error handling
- Change the encoding interface: encode() and decode() instead huffman::encode_string() and encoded.decoded()

### CODE REVIEW

**Comment by Butt4cak3** [comment]
Could you please provide some explanation for why you think this rework was necessary and why your version is better than the current?

**Comment by antonte** [comment]
Updated commit message and message in PR to emphasize why rework is needed.

**Comment by antonte** [comment]
@Gustorn no problem, I'll rename Node and Codebook to lower case.

**Comment by antonte** [comment]
@Gustorn changes with snake_case are pushed

File: contents/huffman_encoding/code/c++/huffman.cpp [hunk 1]
```diff
@@ -1,268 +1,124 @@
...
-  std::vector<bool> encode(Iter first, Iter last) const;
+  // branch fields
+  std::unique_ptr<node> left;
+  std::unique_ptr<node> right;
 
-  template <typename Iter>
-  std::string decode(Iter first, Iter last) const;
+  // common fields
+  node* parent = nullptr;
+  enum class branch { left, right };
```
Comment (Gustorn) [inline contents/huffman_encoding/code/c++/huffman.cpp]: I don't think storing which branch we're on is necessary.

Comment (antonte) [reply contents/huffman_encoding/code/c++/huffman.cpp]: What do you mean? It used here:             
```C++
compressed_char.push_back(node->branch != node::branch::left);
```
https://github.com/antonte/algorithm-archive/blob/71c0d4295d4a892c91e1da64fdb7feaa9ea56dd3/contents/huffman_encoding/code/c%2B%2B/huffman.cpp#L83

**Comment by leios** [comment]
Hey guys. What's the status on this PR? It looks like the review was completed; however, it was not yet merged into the AAA. Can I dismiss the review @Gustorn and merge?

**Comment by Gathros** [comment]
Another stale pull request.

**Pull request closed without merging**